### PR TITLE
subsys: console: Split tty definitions into tty.h, to form its own API

### PR DESCRIPTION
**Pull request by pfalcon**
This patch proceeds with the separation of older serial console
subsystem into device-independent console subsytem and buffered
serial device ("tty") API.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

### CODE REVIEW

**Comment by pfalcon** [comment]
@nashif : Trivial patch for you. Thanks.

**Comment by nashif** [comment]
adding @galak who is looking at include/*.h and public APIs.

**Comment by pfalcon** [comment]
@galak : Ping about this trivial patch.

File: include/tty.h [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+};
+
+/**
+ * @brief Initialize buffered serial port (classically known as tty).
+ *
+ * "tty" device provides buffered, interrupt-driver access to an
+ * underlying UART device.
+ *
+ * @param tty tty device structure to initialize
+ * @param uart_dev underlying UART device to use (should support
```
Comment (Mierunski) [inline include/tty.h]: As I understand, tty API should cover also other transport ways than UART. This should be common API to support UART, CDC ACM, RTT etc. and allow to handle all serial-like devices in the same manner. Like ttyS, ttyACM, ttyUSB in linux

**Review by Mierunski — CHANGES_REQUESTED** [review]

File: include/tty.h (hunk 1 above)
Comment (pfalcon) [reply include/tty.h]: > As I understand, tty API should cover also other transport ways than UART.

Whether that's the case or not is still subject to discussion (ongoing at https://github.com/zephyrproject-rtos/zephyr/issues/10512).

The purpose of this patch is very simple (trivial, indeed, as the label suggests) - to move some arguably misplaced declarations into a header more suitable for them. Per the definition of Trivial-labeled patch, there're no changes to semantics or behavior.

Comment (Mierunski) [reply include/tty.h]: Ok then

**Review by Mierunski — APPROVED** [review]

**Comment by pfalcon** [comment]
@galak: Ping, would like to have steady gradual progress with tty/console API elaboration, and further work is blocked by this trivial patch.

**Comment by pfalcon** [comment]
@galak: You seem to be bust with other things so far, while this patch was intended just as an easy, small step towards elaboration of console API. It was also discussed in https://github.com/zephyrproject-rtos/zephyr/issues/10512, so should be easy to review and merge. It's also holding up further work on console. Let me unassign you and try to get it reviewed, or otherwise I'll just close it.

**Comment by pfalcon** [comment]
@pkral78: Re: discussion in https://github.com/zephyrproject-rtos/zephyr/issues/10512, this is a patch a "critical-step" patch to move in the direction of supporting RTT, etc. Would appreciate help with the review.

**Comment by pfalcon** [comment]
@carlescufi, @nordic-krch, @jarz-nordic: Please review.

**Review by jakub-uC — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Comment by carlescufi** [comment]
@pfalcon I am fine with this patch. Will wait for @galak

**Review by nordic-krch — APPROVED** [review]

**Comment by pfalcon** [comment]
@carlescufi, Fairly speaking, I don't know why @nashif assigned it to @galak, he indeed does a lot of cleanup in includes, but mostly regarding lower-level machine APIs as far as I can tell. I'd say it rather go your way, because related to console stuff and you'd be more in loop on it, as you presented that topic to TSC, etc.

But sure, let's wait for @galak.


Thanks for reviews everyone!

**Comment by pkral78** [comment]
> @pkral78: Re: discussion in #10512, this is a patch a "critical-step" patch to move in the direction of supporting RTT, etc. Would appreciate help with the review.

LGTM, I stay tuned

**Comment by pfalcon** [comment]
@carlescufi, @nashif: Can we merge this well-approved PR?

**Comment by pfalcon** [comment]
@nashif: Can we please merge this?

Because in logging discussion, there's again popping up discussion of shell vs logging vs console, and I guess one of the reply later will be "we don't use console because you didn't finish it!" But I can't proceed with it because trivial patches like this are held in limbo for extended time.

**Pull request merged**